dbms/src/DataStreams/WindowTransformAction.cpp

```diff
@@ -1294,10 +1300,58 @@ void WindowTransformAction::writeOutCurrentRow()
     }
 }
 
+void WindowTransformAction::writeBatchResult()
+{
+    assert(partition_ended);
+    assert(current_row < partition_end);
+
+    size_t insert_row_num = 0;
+    const auto & block = blockAt(current_row);
+    RowNumber tmp_row;
+    if (current_row.block != partition_end.block)
+    {
+        insert_row_num = block.rows - current_row.row;
+        tmp_row = current_row;
+        tmp_row.block += 1;
+        tmp_row.row = 0;
+    }
+    else
+    {
+        insert_row_num = partition_end.row - current_row.row;
+        tmp_row = partition_end;
+    }
+
+    // As aggregation function and window function can not occur in one window at the same time
+    // Only batch inserting results for aggregation function is ok
+    for (auto & ws : aggregation_workspaces)
+    {
+        IColumn * result_column = block.output_columns[ws.idx].get();
+        const auto * agg_func = ws.aggregate_function.get();
+        auto * buf = ws.aggregate_function_state.data();
+        agg_func->insertBatchResultInto(buf, *result_column, insert_row_num, nullptr);
+    }
+
+    current_row = tmp_row;
+}
+
+
 Block WindowTransformAction::tryGetOutputBlock()
 {
     // first try calculate the result based on current data
-    tryCalculate();
+    if (window_description.need_decrease)
```

Comment:
if need_decrease is true, has_shortcut should always be false?

Reply:
> if need_decrease is true, has_shortcut should always be false?

yes

Reply:
then why need check `has_shortcut` in L1340

Reply:
> then why need check `has_shortcut` in L1340

deleted